In [187]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import LeaveOneOut
from sklearn.model_selection import cross_val_predict
from sklearn.model_selection import KFold
from sklearn.utils import resample
from sklearn.metrics import r2_score

In [188]:
communities = pd.read_csv('communities.data', header=None, na_values ="?")

In [189]:
communities.head()

0     1        2                    3    4     5     6     7     8     9    \
0    8   NaN      NaN         Lakewoodcity    1  0.19  0.33  0.02  0.90  0.12   
1   53   NaN      NaN          Tukwilacity    1  0.00  0.16  0.12  0.74  0.45   
2   24   NaN      NaN         Aberdeentown    1  0.00  0.42  0.49  0.56  0.17   
3   34   5.0  81440.0  Willingborotownship    1  0.04  0.77  1.00  0.08  0.12   
4   42  95.0   6096.0    Bethlehemtownship    1  0.01  0.55  0.02  0.95  0.09   

   ...   118   119   120   121   122  123  124   125   126   127  
0  ...  0.12  0.26  0.20  0.06  0.04  0.9  0.5  0.32  0.14  0.20  
1  ...  0.02  0.12  0.45   NaN   NaN  NaN  NaN  0.00   NaN  0.67  
2  ...  0.01  0.21  0.02   NaN   NaN  NaN  NaN  0.00   NaN  0.43  
3  ...  0.02  0.39  0.28   NaN   NaN  NaN  NaN  0.00   NaN  0.12  
4  ...  0.04  0.09  0.02   NaN   NaN  NaN  NaN  0.00   NaN  0.03  

[5 rows x 128 columns]

In [190]:
communities.columns = ['state', 'county', 'community', 'communityname', 'fold', 'population',
    'householdsize', 'racepctblack', 'racePctWhite', 'racePctAsian',
    'racePctHisp', 'agePct12t21', 'agePct12t29', 'agePct16t24', 'agePct65up',
    'numbUrban', 'pctUrban', 'medIncome', 'pctWWage', 'pctWFarmSelf',
    'pctWInvInc', 'pctWSocSec', 'pctWPubAsst', 'pctWRetire', 'medFamInc',
    'perCapInc', 'whitePerCap', 'blackPerCap', 'indianPerCap', 'AsianPerCap',
    'OtherPerCap', 'HispPerCap', 'NumUnderPov', 'PctPopUnderPov',
    'PctLess9thGrade', 'PctNotHSGrad', 'PctBSorMore', 'PctUnemployed',
    'PctEmploy', 'PctEmplManu', 'PctEmplProfServ', 'PctOccupManu',
    'PctOccupMgmtProf', 'MalePctDivorce', 'MalePctNevMarr', 'FemalePctDiv',
    'TotalPctDiv', 'PersPerFam', 'PctFam2Par', 'PctKids2Par',
    'PctYoungKids2Par', 'PctTeen2Par', 'PctWorkMomYoungKids', 'PctWorkMom',
    'NumIlleg', 'PctIlleg', 'NumImmig', 'PctImmigRecent', 'PctImmigRec5',
    'PctImmigRec8', 'PctImmigRec10', 'PctRecentImmig', 'PctRecImmig5',
    'PctRecImmig8', 'PctRecImmig10', 'PctSpeakEnglOnly', 'PctNotSpeakEnglWell',
    'PctLargHouseFam', 'PctLargHouseOccup', 'PersPerOccupHous',
    'PersPerOwnOccHous', 'PersPerRentOccHous', 'PctPersOwnOccup',
    'PctPersDenseHous', 'PctHousLess3BR', 'MedNumBR', 'HousVacant',
    'PctHousOccup', 'PctHousOwnOcc', 'PctVacantBoarded', 'PctVacMore6Mos',
    'MedYrHousBuilt', 'PctHousNoPhone', 'PctWOFullPlumb', 'OwnOccLowQuart',
    'OwnOccMedVal', 'OwnOccHiQuart', 'RentLowQ', 'RentMedian', 'RentHighQ',
    'MedRent', 'MedRentPctHousInc', 'MedOwnCostPctInc', 'MedOwnCostPctIncNoMtg',
    'NumInShelters', 'NumStreet', 'PctForeignBorn', 'PctBornSameState',
    'PctSameHouse85', 'PctSameCity85', 'PctSameState85', 'LemasSwornFT',
    'LemasSwFTPerPop', 'LemasSwFTFieldOps', 'LemasSwFTFieldPerPop',
    'LemasTotalReq', 'LemasTotReqPerPop', 'PolicReqPerOffic', 'PolicPerPop',
    'RacialMatchCommPol', 'PctPolicWhite', 'PctPolicBlack', 'PctPolicHisp',
    'PctPolicAsian', 'PctPolicMinor', 'OfficAssgnDrugUnits',
    'NumKindsDrugsSeiz', 'PolicAveOTWorked', 'LandArea', 'PopDens',
    'PctUsePubTrans', 'PolicCars', 'PolicOperBudg', 'LemasPctPolicOnPatr',
    'LemasGangUnitDeploy', 'LemasPctOfficDrugUn', 'PolicBudgPerPop',
    'ViolentCrimesPerPop'
  ]

In [191]:
communities.head()

state  county  community        communityname  fold  population  \
0      8     NaN        NaN         Lakewoodcity     1        0.19   
1     53     NaN        NaN          Tukwilacity     1        0.00   
2     24     NaN        NaN         Aberdeentown     1        0.00   
3     34     5.0    81440.0  Willingborotownship     1        0.04   
4     42    95.0     6096.0    Bethlehemtownship     1        0.01   

   householdsize  racepctblack  racePctWhite  racePctAsian  ...  LandArea  \
0           0.33          0.02          0.90          0.12  ...      0.12   
1           0.16          0.12          0.74          0.45  ...      0.02   
2           0.42          0.49          0.56          0.17  ...      0.01   
3           0.77          1.00          0.08          0.12  ...      0.02   
4           0.55          0.02          0.95          0.09  ...      0.04   

   PopDens  PctUsePubTrans  PolicCars  PolicOperBudg  LemasPctPolicOnPatr  \
0     0.26            0.20       0.06           0.04                  0.9   
1     0.12            0.45        NaN            NaN                  NaN   
2     0.21            0.02        NaN            NaN                  NaN   
3     0.39            0.28        NaN            NaN                  NaN   
4     0.09            0.02        NaN            NaN                  NaN   

   LemasGangUnitDeploy  LemasPctOfficDrugUn  PolicBudgPerPop  \
0                  0.5                 0.32             0.14   
1                  NaN                 0.00              NaN   
2                  NaN                 0.00              NaN   
3                  NaN                 0.00              NaN   
4                  NaN                 0.00              NaN   

   ViolentCrimesPerPop  
0                 0.20  
1                 0.67  
2                 0.43  
3                 0.12  
4                 0.03  

[5 rows x 128 columns]

In [192]:
Y = communities['ViolentCrimesPerPop']

In [193]:
communities = communities.drop(communities.columns[:5], axis=1)

In [194]:
communities = communities.drop(['ViolentCrimesPerPop'], axis=1)

In [195]:
communities.shape

(1994, 122)

In [196]:
X = communities

In [197]:
X.isnull().values.any()

True

In [198]:
x_train, x_test, y_train, y_test = train_test_split( X, Y, test_size = 0.1, random_state = 13)

In [199]:
x_train.shape

(1794, 122)

In [200]:
medians = x_train.median()
for column in x_train.columns:
    x_train[column].fillna(medians[column], inplace=True)
for column in x_test.columns:
    x_test[column].fillna(medians[column], inplace=True)

### Yes will need to scale the data as we will be computing a linear model with regularization. Unscaled data does not converge fast enough for linear model.

In [201]:
scaler = StandardScaler()
scaler.fit(x_train)

StandardScaler()

In [202]:
X_train = pd.DataFrame(scaler.transform(x_train),columns = x_train.columns)
X_test = pd.DataFrame(scaler.transform(x_test), columns = x_test.columns)

In [203]:
X_train

population  householdsize  racepctblack  racePctWhite  racePctAsian  \
0       0.339896      -0.810953     -0.431946      0.352570     -0.545257   
1      -0.137173       1.082111     -0.632836     -2.289602      4.072706   
2      -0.296196      -1.421619      0.813569     -0.844664      1.908036   
3      -0.296196      -0.566687      3.304600     -2.743726     -0.641464   
4      -0.455219      -0.322421     -0.512302      0.847978     -0.689568   
...          ...            ...           ...           ...           ...   
1789   -0.375707      -0.933087      2.621576     -1.711627     -0.689568   
1790   -0.296196      -0.444554     -0.673014      0.930546     -0.545257   
1791    0.180873      -0.444554     -0.391768      0.393854     -0.545257   
1792   -0.296196       0.043979     -0.673014      0.806694      0.031989   
1793    0.101362      -0.505620      0.251078      0.022299     -0.160426   

      racePctHisp  agePct12t21  agePct12t29  agePct16t24  agePct65up  ...  \
0       -0.064188    -0.536762    -0.029822    -0.218032    0.641726  ...   
1        2.247422    -0.150112     0.247379     0.081424   -0.022333  ...   
2       -0.321033    -0.278995     2.049185     1.458919   -1.405789  ...   
3       -0.577878     0.236537    -0.307022    -0.397705    1.029094  ...   
4       -0.577878    -0.343437    -0.861424    -0.577378    0.697065  ...   
...           ...          ...          ...          ...         ...  ...   
1789    -0.577878    -0.278995    -0.584223    -0.337814    1.305785  ...   
1790    -0.577878    -0.278995    -0.307022    -0.218032    0.254359  ...   
1791     0.449503    -0.601203    -0.099122    -0.397705    0.254359  ...   
1792    -0.577878    -0.085671    -0.722824    -0.517487   -0.575715  ...   
1793    -0.363841    -0.665645     0.039479    -0.098249    0.309697  ...   

      PolicAveOTWorked  LandArea   PopDens  PctUsePubTrans  PolicCars  \
0            -0.846782 -0.414848  2.258614        0.082841  -0.369049   
1            -0.094845 -0.507243  2.501418        0.479151  -0.144731   
2            -0.094845 -0.507243  0.316186        1.624048  -0.144731   
3            -0.094845 -0.230059 -0.509346       -0.533642  -0.144731   
4            -0.094845  0.231913 -0.994953       -0.489608  -0.144731   
...                ...       ...       ...             ...        ...   
1789         -0.094845 -0.414848 -0.460785       -0.709780  -0.144731   
1790         -0.094845  0.971069 -0.994953       -0.577677  -0.144731   
1791          0.549672  0.047124 -0.120860       -0.533642  -0.817684   
1792         -0.094845 -0.137665 -0.655028       -0.621711  -0.144731   
1793         -2.028396 -0.322454  0.898914        0.126875  -0.817684   

      PolicOperBudg  LemasPctPolicOnPatr  LemasGangUnitDeploy  \
0         -0.460508             1.092844             0.054322   
1         -0.124538             0.090967             0.054322   
2         -0.124538             0.090967             0.054322   
3         -0.124538             0.090967             0.054322   
4         -0.124538             0.090967             0.054322   
...             ...                  ...                  ...   
1789      -0.124538             0.090967             0.054322   
1790      -0.124538             0.090967             0.054322   
1791      -0.460508            -0.465631             3.099763   
1792      -0.124538             0.090967             0.054322   
1793      -0.460508             2.094721            -2.991119   

      LemasPctOfficDrugUn  PolicBudgPerPop  
0                3.742496        -0.849501  
1               -0.394821        -0.106374  
2               -0.394821        -0.106374  
3               -0.394821        -0.106374  
4               -0.394821        -0.106374  
...                   ...              ...  
1789            -0.394821        -0.106374  
1790            -0.394821        -0.106374  
1791             1.425598        -0.700876  
1792            -0.394821        -0.106374  
179

In [230]:
correlations = [
    np.abs(np.corrcoef(y_train, X_train.iloc[:,i])[0][1]) for i in range(X_train.shape[1])
]
corr_df = pd.DataFrame({'variables': X_train.columns, 'corr': correlations})
corr_df = corr_df.reindex(corr_df['corr'].sort_values(ascending=False).index)

In [231]:
corr_df

variables      corr
44           PctKids2Par  0.734404
50              PctIlleg  0.731454
43            PctFam2Par  0.701110
3           racePctWhite  0.678336
45      PctYoungKids2Par  0.661423
..                   ...       ...
111    NumKindsDrugsSeiz  0.033370
1          householdsize  0.031036
95        PctSameState85  0.023219
75        PctVacMore6Mos  0.012351
119  LemasGangUnitDeploy  0.000143

[122 rows x 2 columns]

### 5 features with strongest correlation: PctKids2Par, Pctllleg, PctFam2Par, racePctWhite, PctYoungKids2Par.

### 5 features with weakest correlation: PNumKindsDrugsSeiz, householdsize, PctSameState85, PctVacMore6Mos, LemasGangUnitDeploy.

In [232]:
alphas = [1e-3, 1e-2, 1e-1, 1, 10]

In [233]:
cv = KFold(n_splits = 10, random_state = 13, shuffle = True)

In [234]:
from sklearn.linear_model import LassoCV
lasso_cv = LassoCV(alphas=alphas, cv=cv)

In [235]:
lasso_cv.fit(X_train, y_train)

LassoCV(alphas=[0.001, 0.01, 0.1, 1, 10],
        cv=KFold(n_splits=10, random_state=13, shuffle=True))

In [236]:
print(f"The best regularization constant is: {lasso_cv.alpha_}")

The best regularization constant is: 0.001


In [237]:
lasso_cv.score(X_train, y_train)

0.677991958206577

### R2 returned by the selected model is 0.677991958206577 which means that 67.8% of the variance in the target variable is explained by the model. A higher R2 indicates a better fit of the model to the data.

In [238]:
print(f"Number of Selected features: {np.sum(lasso_cv.coef_ != 0)}")

Number of Selected features: 64


In [239]:
print(f"Selected features: {X_train.columns[lasso_cv.coef_ != 0]}")

Selected features: Index(['householdsize', 'racepctblack', 'racePctWhite', 'agePct12t29',
       'numbUrban', 'pctUrban', 'pctWWage', 'pctWFarmSelf', 'pctWInvInc',
       'pctWSocSec', 'pctWRetire', 'whitePerCap', 'blackPerCap',
       'indianPerCap', 'AsianPerCap', 'OtherPerCap', 'HispPerCap',
       'PctPopUnderPov', 'PctLess9thGrade', 'PctUnemployed', 'PctEmploy',
       'PctEmplManu', 'MalePctDivorce', 'MalePctNevMarr', 'PctKids2Par',
       'PctYoungKids2Par', 'PctWorkMom', 'NumIlleg', 'PctIlleg', 'NumImmig',
       'PctImmigRec5', 'PctRecImmig8', 'PctNotSpeakEnglWell',
       'PersPerOccupHous', 'PersPerOwnOccHous', 'PctPersOwnOccup',
       'PctPersDenseHous', 'PctHousLess3BR', 'HousVacant', 'PctHousOccup',
       'PctVacantBoarded', 'PctVacMore6Mos', 'RentLowQ', 'MedRent',
       'MedRentPctHousInc', 'MedOwnCostPctIncNoMtg', 'NumInShelters',
       'NumStreet', 'PctForeignBorn', 'PctBornSameState', 'PctSameCity85',
       'LemasSwFTFieldPerPop', 'LemasTotalReq', 'PolicReqPerOff

### Lasso minimizes the following criterion: J(β) = RSS + α * Σ|β_i|. RSS is the residual sum of squares and α is the regularization strength.
### RSS is differentiable with respect to the model's coefficients (β) because it involves basic operations that are all differentiable.
### The L1 regularization term Σ|β_i| is not differentiable at zero because it contains absolute value functions. The derivative is not defined at zero. However, the L1 regularization term is subdifferentiable, meaning it has subgradients at zero. Subgradients are used to generalize the concept of derivatives for non-differentiable functions.
### As a result, the entire criterion J(β) is not differentiable at zero due to the L1 regularization term. However, it is differentiable everywhere else.
### The closed form solution for Lasso may involve gradient descent or coordinate descent, and it's typically implemented using optimization algorithms to iteratively update the coefficients to minimize the criterion ensuring that the coefficients reach the optimal values.

In [240]:
from sklearn.linear_model import RidgeCV
ridge_cv = RidgeCV(alphas=alphas, cv=cv)
ridge_cv.fit(X_train, y_train)

RidgeCV(alphas=[0.001, 0.01, 0.1, 1, 10],
        cv=KFold(n_splits=10, random_state=13, shuffle=True))

In [241]:
print(f"The best alpha is: {ridge_cv.alpha_}")

The best alpha is: 10.0


In [242]:
print(f"The R2 value is: {ridge_cv.score(X_train, y_train)}")

The R2 value is: 0.6936307407214163


### Ridge minimizes the following criterion: J(β) = RSS + α * Σ(β_i^2).  RSS is the residual sum of squares, α is the regularization strength, and β_i are the model coefficients.

### The Ridge regression criterion is differentiable in its parameters. The derivative of the Ridge criterion with respect to the coefficients β_i is dJ(β)/dβ_i = -2 * (y_i - X_iβ) + 2αβ_i
### The closed form solution of the estimator of the parameter vector (β) in Ridge regression is: β^Ridge = (X^T X + αI)^(-1) X^T y where X is the feature matrix, y is the target vector, α is the regularization strength, and I is the identity matrix.

### L1 Lasso is simpler in terms of having less features (coefficients shrunk to zero). L2 provides a smoother regularization with a slightly higher R2 value. Both models fit about the same and L1 is simpler with less features involved.

In [243]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import RandomizedSearchCV

In [244]:
param_dist = {
    'min_samples_leaf': np.arange(1, 61),
    'max_features': np.linspace(0.1, 1.0, 10),
    'max_depth': np.arange(6, 21),
}

In [245]:
rf = RandomForestRegressor(n_estimators=100)

In [246]:
random_search = RandomizedSearchCV(estimator=rf, param_distributions=param_dist, n_iter=100, n_jobs=-1, scoring='r2')

In [247]:
random_search.fit(X_train, y_train)

RandomizedSearchCV(estimator=RandomForestRegressor(), n_iter=100, n_jobs=-1,
                   param_distributions={'max_depth': array([ 6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]),
                                        'max_features': array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. ]),
                                        'min_samples_leaf': array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58, 59, 60])},
                   scoring='r2')

In [248]:
print(f"Best hyperparameter values are: {random_search.best_params_}")

Best hyperparameter values are: {'min_samples_leaf': 3, 'max_features': 0.4, 'max_depth': 11}


### The best hyperparameter values These values indicate which combinations of hyperparameters resulted in the best model performance.

In [249]:
best_model = random_search.best_estimator_
best_r2 = best_model.score(X_train, y_train)
print("R2 score of the best model:", best_r2)

R2 score of the best model: 0.904330878070565


### 7 1). Rsquared score is used to evaluate models. The random forest model with hyperparameter values of 'min_samples_leaf': 3, 'max_features': 0.4 'max_depth': 11 is the best fit model with the highest Rsquared value.
### 2). Out of all 3 cv-selected models, Lasso is the best in terms of simplicity and interpretability. Linear models are easier to interpret than Random Forest as it is clear to see feature contributions to the target by looking at the coefficients. Lasso is simpler compared to the other two as it has less features involved in the model.
### 3). I would choose the Random Forest model as the it provides a significant better fit (aka. highest Rsquared value). 
### 4). Yes the top 5 features and all other features are retained in the Random Forest model as it does not do feature selection during the modeling process. If feature selection is needed, it can be done before the model fitting through univariate feature selection.

In [250]:
hyperparameters = {
    'min_samples_leaf': 3,
    'max_features': 0.4,
    'max_depth': 12
}

In [251]:
rf_test = RandomForestRegressor(**hyperparameters)
rf_test.fit(X_test, y_test)

RandomForestRegressor(max_depth=12, max_features=0.4, min_samples_leaf=3)

In [252]:
rf_test.score(X_test, y_test)

0.9234591964039427

### The estimated R2 is 0.9234591964039427 which relatively optimistimal as it's close to the R2 from training set. 

In [253]:
for i in range(100):
    r2_values = []
    random_seed = np.random.randint(1, 1000)
    x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=random_seed)
    
    medians = x_train.median()
    for column in x_train.columns:
        x_train[column].fillna(medians[column], inplace=True)
    for column in x_test.columns:
        x_test[column].fillna(medians[column], inplace=True)

    model = RandomForestRegressor(**hyperparameters)
    model.fit(x_train, y_train)

    r2 = model.score(x_test, y_test)

    r2_values.append(r2)

In [254]:
mean_r2 = np.mean(r2_values)
std_r2 = np.std(r2_values)
lower_bound = mean_r2 - 1.96 * (std_r2 / np.sqrt(100))
upper_bound = mean_r2 + 1.96 * (std_r2 / np.sqrt(100))

In [255]:
mean_r2

0.6668436978535583

In [256]:
lower_bound, upper_bound

(0.6668436978535583, 0.6668436978535583)

### The mean is 0.6668436978535583. The 95% confidence interval is (0.6668436978535583, 0.6668436978535583) for the true value of R2. 